lib/softoken/kbkdf.c

```diff
@@ -0,0 +1,427 @@
+
+        ret = param_validate(params->pDataParams + offset);
+        if (ret != CKR_OK) {
+            PR_ASSERT(ret == CKR_OK);
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+    }
+
+    /* We have to have at least one iteration variable parameter. */
+    if (param_type_count[CK_SP800_108_ITERATION_VARIABLE] == 0) {
+        PR_ASSERT(PR_FALSE);
+        return CKR_MECHANISM_PARAM_INVALID;
+    }
+
+    /* We can't have any counters parameters. */
+    if (param_type_count[CK_SP800_108_OPTIONAL_COUNTER] != 0) {
+        PR_ASSERT(PR_FALSE);
+        return CKR_MECHANISM_PARAM_INVALID;
+    }
+
+    /* We have to have one DKM length parameter. Otherwise the size and
+     * endianness of the length parameter is undefined, even without
+     * additional derived keys. */
+    if (param_type_count[CK_SP800_108_DKM_LENGTH] != 1) {
+        PR_ASSERT(PR_FALSE);
+        return CKR_MECHANISM_PARAM_INVALID;
+    }
+
+    /* Validate basic assumptions about derived keys:
+     *      NULL <-> ulAdditionalDerivedKeys > 0
+     */
+    if ((params->ulAdditionalDerivedKeys > 0) != (params->pAdditionalDerivedKeys != NULL)) {
+        PR_ASSERT(PR_FALSE);
+        return CKR_MECHANISM_PARAM_INVALID;
+    }
+
+    /* Validate each derived key. */
+    for (offset = 0; offset < params->ulAdditionalDerivedKeys; offset++) {
+        ret = derived_validate(params->pAdditionalDerivedKeys + offset);
+        if (ret != CKR_OK) {
+            PR_ASSERT(ret == CKR_OK);
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+    }
+
+    return CKR_OK;
+}
+
+/* [ section: parameter helpers ] */
+static CK_RV kbkdf_CalculateParameters() {
+    return CKR_OK;
+}
+
+/* [ section: PRF helpers ] */
+
+static CK_RV kbkdf_PrfInit(CK_MECHANISM_TYPE mech, SFTKObject *key, MACCtx *ctx) {
+    /* TODO : Implement a way of creating all required context for a call to NSC_Sign. */
+
+    /* For now, implement this with HMAC / CMAC directly. */
+
+    SFTKAttribute *keyval;
+    const SECHashObject *hashObj;
+    PRBool isFIPS = (key->slot->slotID == FIPS_SLOT_ID);
+    CK_RV ret = CKR_OK;
+
+    keyval = sftk_FindAttribute(key, CKA_VALUE);
+    if (keyval == NULL) {
+        ret = CKR_KEY_SIZE_RANGE;
+        goto failure;
+    }
+
+    switch (mech) {
+    case CKM_SHA_1_HMAC:
+        hashObj = HASH_GetRawHashObject(HASH_AlgSHA1);
+        ctx->prf_size = SHA1_LENGTH;
+        goto hmac;
+    case CKM_SHA224_HMAC:
+        hashObj = HASH_GetRawHashObject(HASH_AlgSHA224);
+        ctx->prf_size = SHA224_LENGTH;
+        goto hmac;
+    case CKM_SHA256_HMAC:
+        hashObj = HASH_GetRawHashObject(HASH_AlgSHA256);
+        ctx->prf_size = SHA256_LENGTH;
+        goto hmac;
+    case CKM_SHA384_HMAC:
+        hashObj = HASH_GetRawHashObject(HASH_AlgSHA384);
+        ctx->prf_size = SHA384_LENGTH;
+        goto hmac;
+    case CKM_SHA512_HMAC:
+        hashObj = HASH_GetRawHashObject(HASH_AlgSHA512);
+        ctx->prf_size = SHA512_LENGTH;
+        goto hmac;
```

Comment:
@rjrelyea I was actually hoping to avoid this all together but I wasn't yet quite sure how so I went with the way I knew would work. :)

My idea was to use the existing PKCS#11 layer to handle this for us. Rather than holding the `MACCtx ctx` object everywhere (and having to handle the different HMAC/CMAC initializations/update/finalization code), keep only a `CK_MECHANISM_PTR pMechanism` (and at most a `CK_SESSION_HANDLE hSession`) to be able to call `NSC_SignInit`/`NSC_SignUpdate`/`NSC_Sign`.

The questions I had left unresolved were the blockers:

 - Can I reuse `hSession` from the parent (NIST 800-108 KBKDF) context or do I have to create a new one for HMAC/CMAC?
 - How do I construct `CK_MECHANISM_PTR`?

Anyhow, I went with this current approach so I could work towards having a valid Counter implementation while I figured the answers to those questions out. 

Thoughts?

Reply:
(And to answer the question: `ike` has a `sftk_map_hmac_to_hash` function we could probably reuse once we add a header definition).

Reply:
So that's a bit tricky. sftk_InitCBCMac() sort of does this. The problem is technically the session could already have a Sign session going on it (not common, but possible), so reusing that session would be iffy. We could create a new session. You'll also need to create specific keys (since the input key is likely modified before using in the hmac). sftk_InitCBCMac does this as well (It creates a temp key used for the underlying AES operation), That key is  freed before the Init returns. We have to remember to free any of these things we create in the operation.

In ike, I created a generic mac interface that kindof handled both hmac (which is pretty easy) and aes_xcbc. Maybe we should create a completely generic version that handles all the macing. That might be too much for this patch, but it might make a good future bug.

bob